src/coreclr/jit/ssabuilder.cpp

```diff
@@ -1505,3 +1361,388 @@ void Compiler::JitTestCheckSSA()
     }
 }
 #endif // DEBUG
+
+class IncrementalSsaBuilder
+{
+    Compiler*                   m_comp;
+    unsigned                    m_lclNum;
+    ArrayStack<UseDefLocation>& m_defs;
+    ArrayStack<UseDefLocation>& m_uses;
+    BitVecTraits                m_poTraits;
+    BitVec                      m_defBlocks;
+    BitVec                      m_iteratedDominanceFrontiers;
+
+    UseDefLocation FindOrCreateReachingDef(const UseDefLocation& use);
+    bool           FindReachingDefInBlock(const UseDefLocation& use, BasicBlock* block, UseDefLocation* def);
+    bool           FindReachingDefInSameStatement(const UseDefLocation& use, UseDefLocation* def);
+    Statement*     LatestStatement(Statement* stmt1, Statement* stmt2);
+public:
+    IncrementalSsaBuilder(Compiler*                   comp,
+                          unsigned                    lclNum,
+                          ArrayStack<UseDefLocation>& defs,
+                          ArrayStack<UseDefLocation>& uses)
+        : m_comp(comp)
+        , m_lclNum(lclNum)
+        , m_defs(defs)
+        , m_uses(uses)
+        , m_poTraits(comp->m_dfsTree->PostOrderTraits())
+        , m_defBlocks(BitVecOps::MakeEmpty(&m_poTraits))
+        , m_iteratedDominanceFrontiers(BitVecOps::MakeEmpty(&m_poTraits))
+    {
+    }
+
+    void Insert();
+};
+
+//------------------------------------------------------------------------
+// FindOrCreateReachingDef: Given a use indicated by a block and potentially a
+// statement and tree, find the reaching definition for it, potentially
+// creating it if the reaching definition is a phi that has not been created
+// yet.
+//
+// Parameters:
+//   use - The use. The block must be non-null. The statement and tree can be
+//         null, meaning that the use is happening after the last statement in the
+//         block.
+//
+// Returns:
+//   Location of a definition node that is the reaching def.
+//
+UseDefLocation IncrementalSsaBuilder::FindOrCreateReachingDef(const UseDefLocation& use)
+{
+    for (BasicBlock* dom = use.Block; dom != nullptr; dom = dom->bbIDom)
+    {
+        UseDefLocation reachingDef;
+        if (BitVecOps::IsMember(&m_poTraits, m_defBlocks, dom->bbPostorderNum) &&
+            FindReachingDefInBlock(use, dom, &reachingDef))
+        {
+            return reachingDef;
+        }
+
+        if (BitVecOps::IsMember(&m_poTraits, m_iteratedDominanceFrontiers, dom->bbPostorderNum))
```

Comment:
Seems like if you are worried about worst-case costs and can be allowed to fail to put a local into SSA, you could see if there is any particularly large IDF and just punt if there is one (maybe?).

Reply:
Also maybe only handle the two-def case (not sure we ever really see more than two defs for a CSE)?

Reply:
> Seems like if you are worried about worst-case costs and can be allowed to fail to put a local into SSA, you could see if there is any particularly large IDF and just punt if there is one (maybe?).

Yeah that's a good point, the number of recursive calls is bounded by the IDF size. I just dislike making the updater partial in this way...

Reply:
I went with your suggestion and skip the insertion into SSA if the IDF size is > 100. However, I didn't see a need to restrict the number of defs we allow in the inserter.